File: src/main/java/com/tibagni/logviewer/updates/UpdateManager.java
package com.tibagni.logviewer.updates;

import com.tibagni.logviewer.AppInfo;
import com.tibagni.logviewer.logger.Logger;
import com.tibagni.logviewer.util.SwingUtils;
import org.apache.commons.io.IOUtils;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.*;
import java.net.URL;
import java.nio.charset.Charset;

public class UpdateManager {
  private final double CURRENT_VERSION;

  private UpdateListener listener;

  public UpdateManager(UpdateListener listener) {
    CURRENT_VERSION = AppInfo.getCurrentVersionNumber();
    this.listener = listener;
  }

  public void checkForUpdates() {
    // If we fail to read the current version for some reason, do not proceed
    if (CURRENT_VERSION < 0) {
      Logger.error("Not a valid current version. Do not check for updates");
      return;
    }

    Logger.debug("Start checking for updates...");
    SwingUtils.doAsync(
        () ->  getLatestReleaseInfo(),
        latest -> notifyIfUpdateAvailable(latest, listener),
        tr -> {
          Logger.error("Not possible to get latest release info", tr);
          listener.onFailedToCheckForUpdate(tr);
        }
    );
  }

  private void notifyIfUpdateAvailable(ReleaseInfo latest, UpdateListener listener) {
    if (latest.getVersion() > CURRENT_VERSION) {
      Logger.debug("New version available: " + latest.getVersionName());
      listener.onUpdateFound(latest);
    } else {
      Logger.debug("LogViewer is up to date on version " + CURRENT_VERSION);
      listener.onUpToDate();
    }
  }

  private ReleaseInfo getLatestReleaseInfo() throws InvalidReleaseException {
    try {
      URL url = new URL(AppInfo.LATEST_RELEASE_URL);
      JSONObject jsonResult = new JSONObject(IOUtils.toString(url, Charset.forName("UTF-8")));
      return new ReleaseInfo(jsonResult);
    } catch (IOException | JSONException e) {
      throw new InvalidReleaseException(e);
    }
  }

  public interface UpdateListener {
    void onUpdateFound(ReleaseInfo newRelease);
    void onUpToDate();
    void onFailedToCheckForUpdate(Throwable tr);
  }
}

File: src/main/java/com/tibagni/logviewer/updates/UpdateManager.kt
package com.tibagni.logviewer.updates

import com.tibagni.logviewer.AppInfo
import com.tibagni.logviewer.logger.Logger
import com.tibagni.logviewer.util.SwingUtils
import org.apache.commons.io.IOUtils
import org.json.JSONObject
import java.net.URL
import java.nio.charset.Charset

class UpdateManager(private val listener: UpdateListener) {
  interface UpdateListener {
    fun onUpdateFound(newRelease: ReleaseInfo)
    fun onUpToDate()
    fun onFailedToCheckForUpdate(tr: Throwable)
  }

  fun checkForUpdates() {
    // If we fail to read the current version for some reason, do not proceed
    if (AppInfo.currentVersionNumber < 0) {
      Logger.error("Not a valid current version. Do not check for updates")
      return
    }
    Logger.debug("Start checking for updates...")
    SwingUtils.doAsync(
      { latestReleaseInfo },
      { latest: ReleaseInfo -> notifyIfUpdateAvailable(latest, listener) }
    ) { tr: Throwable ->
      Logger.error("Not possible to get latest release info", tr)
      listener.onFailedToCheckForUpdate(tr)
    }
  }

  private fun notifyIfUpdateAvailable(latest: ReleaseInfo, listener: UpdateListener) {
    val currentVersion = AppInfo.currentVersionNumber
    if (latest.version > currentVersion) {
      Logger.debug("New version available: " + latest.versionName)
      listener.onUpdateFound(latest)
    } else {
      Logger.debug("LogViewer is up to date on version $currentVersion")
      listener.onUpToDate()
    }
  }

  @get:Throws(InvalidReleaseException::class)
  private val latestReleaseInfo: ReleaseInfo
    get() = try {
      val url = URL(AppInfo.LATEST_RELEASE_URL)
      val jsonResult = JSONObject(IOUtils.toString(url, Charset.forName("UTF-8")))
      ReleaseInfo(jsonResult)
    } catch (e: Exception) {
      throw InvalidReleaseException(e)
    }
}